feat: implement MacosTable

pithuene opened the pull request:
I implemented a `MacosTable` widget which looks similar to the macOS `NSTableView` (adresses #262 and #199).
The table features ordering by column and row selection.
I have also added some tests for the basic operations.
I am sure there is still a lot to improve, both in code and in UI, but I hope this can serve as a starting point.

This is a screenshot of the table widget running in a process monitor application I am building.
Note that this is running on linux and the font is not the one used on macOS.
![Screenshot_2022-11-18_14-13-43](https://user-images.githubusercontent.com/21110295/202714510-e1898511-51ae-48fd-97db-f3182a07b132.png)

## Pre-launch Checklist

- [x] I have run `dartfmt` on all changed files <!-- THIS IS REQUIRED -->
- [x] I have incremented the package version as appropriate and updated `CHANGELOG.md` with my changes <!-- THIS IS REQUIRED -->
- [ ] I have added/updated relevant documentation <!-- If relevant -->
- [ ] I have run "optimize/organize imports" on all changed files
- [x] I have addressed all analyzer warnings as best I could
<!-- - [ ] I have run `flutter pub publish --dry-run` and addressed any warnings --> <!-- MAINTAINER ONLY -->

pithuene committed:
feat: implement MacosTable

Yetispapa commented:
This is awesome! @GroovinChip, is there something in the way for merging?

stMerlHin commented:
Beautiful, can't wait to see it in the new release.

stMerlHin commented:
Beautiful, can't wait to see it in the new release.

GroovinChip commented on lib/macos_ui.dart:
```diff
@@ -63,6 +63,7 @@ export 'src/selectors/color_well.dart';
 export 'src/selectors/date_picker.dart';
 export 'src/selectors/time_picker.dart';
 export 'src/sheets/macos_sheet.dart';
+export 'src/table/macos_table.dart';
```
I believe a better location for this would be under `/layout`, since it's categorized under "Layout and organization" on the Human Interface Guidelines.

GroovinChip commented on lib/src/table/macos_table_columns.dart:
```diff
@@ -0,0 +1,21 @@
+  final String label;
+  final TableColumnWidth width;
+  final ColumnAlignment alignment;
+  final Widget Function(BuildContext, T) cellBuilder;
```
Please add documentation for all this

GroovinChip commented on lib/src/table/macos_table_datasource.dart:
```diff
@@ -0,0 +1,146 @@
+
+  /// A persistent identifier for this row.
+  ///
+  /// If you select a row and change the data or order,
+  /// the key is how the table knows which row was selected.
```
If you could rewrite this without the use of "you", that would be great - something along the lines of "The key allows the table to know which row has been selected". It's descriptive without addressing the reader directly.

GroovinChip commented on lib/src/table/macos_table_datasource.dart:
```diff
@@ -0,0 +1,146 @@
+  /// If you select a row and change the data or order,
+  /// the key is how the table knows which row was selected.
+  ///
+  /// If you never change the object instances displayed by the table,
+  /// use an `ObjectKey` of the value object.
+  /// Normally, this should be a ValueKey of some unique attribute of the row value.
+  /// If you never even change the order, you could also use a
+  /// ValueKey of the index.
```
Avoid addressing the reader directly

GroovinChip commented on lib/src/table/macos_table_datasource.dart:
```diff
@@ -0,0 +1,146 @@
+  final List<ColumnDefinition<T>> _colDefs;
+  List<ColumnDefinition<T>> get colDefs => _colDefs;
+
+  // By what column and in what direction the table is ordered.
```
```suggestion
  /// The order and direction with which the table can be sorted.
```

GroovinChip commented on lib/src/table/macos_table_datasource.dart:
```diff
@@ -0,0 +1,146 @@
+  List<ColumnDefinition<T>> get colDefs => _colDefs;
+
+  // By what column and in what direction the table is ordered.
+  MacosTableOrder<T>? order;
```
Change this to `tableOrder`.

GroovinChip commented on lib/src/table/macos_table_datasource.dart:
```diff
@@ -0,0 +1,146 @@
+  // By what column and in what direction the table is ordered.
+  MacosTableOrder<T>? order;
+
+  // Callback to update the row count.
```
```suggestion
  /// Callback to update the row count.
```

GroovinChip commented on lib/src/table/macos_table_datasource.dart:
```diff
@@ -0,0 +1,146 @@
+
+  int get rowCount => _rowCount;
+
+  /// The main function through which data is fed into the table.
+  /// You need to maintain some order from index (between 0 and rowCount)
```
```suggestion
  /// The main function through which data is fed into the table.
  ///
  /// You need to maintain some order from index (between 0 and rowCount).
```
Also, can you please explain what you mean by `You need to maintain some order from index (between 0 and rowCount)`?

GroovinChip commented on lib/src/table/macos_table_datasource.dart:
```diff
@@ -0,0 +1,146 @@
+  /// Should either reject the order (by returning `false`),
+  /// or change the order of items returned by [getRowValue] according to
+  /// the given [MacosTableOrder].
+  final bool Function(MacosTableOrder<T>?)? changeOrder;
```
Rename to `changeTableOrder`

GroovinChip commented on lib/src/table/macos_table_datasource.dart:
```diff
@@ -0,0 +1,146 @@
+  // TODO: Allow selecting multiple rows.
+
+  /// The current selection of the table.
+  MacosTableSelection? _selection;
```
Rename to `_tableSelection`.

GroovinChip commented on lib/src/table/macos_table_header.dart:
```diff
@@ -0,0 +1,152 @@
+import 'package:flutter/material.dart';
+import 'package:macos_ui/macos_ui.dart';
+
+class MacosTableHeader<T> extends StatelessWidget {
+  const MacosTableHeader({
+    super.key,
+    required this.colDefs,
+    required this.order,
+    this.columnHeaderClicked,
+  });
+
+  static const double horizontalPadding = 10;
+
+  final List<ColumnDefinition<T>> colDefs;
+  final Function(ColumnDefinition<T>)? columnHeaderClicked;
+
+  final MacosTableOrder<T>? order;
```
Please add documentation.

GroovinChip commented on lib/src/table/macos_table_header.dart:
```diff
@@ -0,0 +1,152 @@
+
+  static const double horizontalPadding = 10;
+
+  final List<ColumnDefinition<T>> colDefs;
```
Please rename to `columnDefinitions`.

GroovinChip commented on lib/src/table/macos_table_order.dart:
```diff
@@ -0,0 +1,24 @@
+import 'package:macos_ui/macos_ui.dart';
+
+enum MacosTableOrderDirection {
+  ascending,
+  descending,
+}
+
+class MacosTableOrder<T> {
+  MacosTableOrder({
+    required this.column,
+    this.direction = MacosTableOrderDirection.descending,
+  });
+
+  ColumnDefinition<T> column;
+  MacosTableOrderDirection direction;
+
+  void reverseDirection() {
```
Please add documentation.

GroovinChip commented on lib/src/table/macos_table_row.dart:
```diff
@@ -0,0 +1,122 @@
+import 'package:macos_ui/macos_ui.dart';
+import 'package:macos_ui/src/table/macos_table_selection.dart';
+
+class MacosTableRow<T> extends StatelessWidget {
+  const MacosTableRow({
+    super.key,
+    required this.index,
+    required this.columnWidths,
+    required this.colDefs,
+    required this.row,
+    required this.rowHeight,
+    required this.data,
+  }) : hasEvenRowHighlight = index % 2 == 1;
+
+  final Map<int, TableColumnWidth> columnWidths;
+  final List<ColumnDefinition<T>> colDefs;
+  final MacosTableValue<T> row;
+
+  final double rowHeight;
+
+  final int index;
+
+  final MacosTableDataSource<T> data;
+
+  final bool hasEvenRowHighlight;
```
Please add documentation.

GroovinChip commented on lib/src/table/macos_table_row.dart:
```diff
@@ -0,0 +1,122 @@
+  }) : hasEvenRowHighlight = index % 2 == 1;
+
+  final Map<int, TableColumnWidth> columnWidths;
+  final List<ColumnDefinition<T>> colDefs;
```
Please rename to `columnDefinitions`.

GroovinChip commented on lib/src/table/macos_table_selection.dart:
```diff
@@ -0,0 +1,23 @@
+import 'package:macos_ui/src/library.dart';
+
+@immutable
+class MacosTableSelection<T> {
+  const MacosTableSelection({
+    required this.key,
+    required this.value,
+  });
+
+  final Key key;
+  final T value;
+}
+
+@immutable
+class MacosTableSelectionChange {
+  const MacosTableSelectionChange(
+    this.oldSelection,
+    this.newSelection,
+  );
+
+  final MacosTableSelection? oldSelection;
+  final MacosTableSelection? newSelection;
```
Please add documentation.

GroovinChip commented on lib/src/table/macos_table_view.dart:
```diff
@@ -0,0 +1,72 @@
+import './macos_table_row.dart';
+import './macos_table_datasource.dart';
+
+/// A scrollable data table with sorting and selection.
```
Please add some more detail to the documentation here. Use the Human Interface Guidelines if you need ideas.

Additionally, please turn this doc into a template and use the macro on the constructor.

GroovinChip commented on lib/src/table/macos_table_view.dart:
```diff
@@ -0,0 +1,72 @@
+
+  /// Optionally override the scrollController.
+  final ScrollController? scrollController;
+  final MacosTableDataSource<T> dataSource;
```
Please add documentation.

GroovinChip reviewed (changes requested):
Hello @pithuene, and thank you for this PR! I'm excited to see a new widget being added, especially this one! 

Please review the feedback I've left and implement the requested changes. Additionally, please update the `CHANGELOG.md` file.

Thanks!

GroovinChip committed:
fix(tests): account for Jan -> Dec & Dec -> Jan

date_picker_test.dart was failing due to not accounting for going from January to December and vice-versa.

pithuene committed:
fix: move table widget to src/layout/table

pithuene committed:
fix: add documentation and clean up some minor things in MacosTable

pithuene commented:
Hey @GroovinChip, thanks a lot for giving me feedback on my pull request!
I tried to implement your requested changes but am not sure what you were asking for in the changelog file.

Thanks again for your time!

GroovinChip commented:
> Hey @GroovinChip, thanks a lot for giving me feedback on my pull request! I tried to implement your requested changes but am not sure what you were asking for in the changelog file.
> 
> Thanks again for your time!

I mean add an entry that describes what the new version adds or changes. If you look at the file you'll understand.

GroovinChip commented on lib/src/layout/table/macos_table_columns.dart:
```diff
@@ -0,0 +1,32 @@
+  final String label;
+
+  /// The width of this column.
+  /// Usually either a [FlexColumnWidth] or a [FixedColumnWidth].
```
```suggestion
  /// The width of this column.
  ///
  /// Usually either a [FlexColumnWidth] or a [FixedColumnWidth].
```

GroovinChip commented on lib/src/layout/table/macos_table_columns.dart:
```diff
@@ -0,0 +1,32 @@
+  final ColumnAlignment alignment;
+
+  /// The builder that creates the content of a cell.
+  /// Receives the current [BuildContext] and the row value.
```
```suggestion
  /// The builder that creates the content of a cell.
  ///
  /// Receives the current [BuildContext] and the row value.
```

GroovinChip commented on lib/src/layout/table/macos_table_datasource.dart:
```diff
@@ -0,0 +1,151 @@
+  final int Function() getRowCount;
+
+  /// The number of rows in the table.
+  /// Limits with which indices `getRow` is called.
```
```suggestion
  /// The number of rows in the table.
  ///
  /// Limits with which indices `getRow` is called.
```

GroovinChip commented on lib/src/layout/table/macos_table_datasource.dart:
```diff
@@ -0,0 +1,151 @@
+  /// Notify the table about new or modified data.
+  ///
+  /// Call this after every data change you want to see,
+  /// if you don't call this function, the new data may not be shown.
```
```suggestion
  /// Must be called after every data change. If it is not called, 
  /// the updated data may not be shown.
  ///
```

GroovinChip commented on lib/src/layout/table/macos_table_datasource.dart:
```diff
@@ -0,0 +1,151 @@
+  /// Normally, this should be a [ValueKey] of some unique attribute
+  /// of the row value.
+  /// If even the order is never changed a [ValueKey] of the index
+  /// would also suffice.
```
```suggestion
  ///
  /// Even if the order is never changed, a [ValueKey] of the index
  /// would also suffice.
```

GroovinChip commented on lib/src/layout/table/macos_table_header.dart:
```diff
@@ -0,0 +1,154 @@
+import 'package:flutter/material.dart';
+import 'package:macos_ui/macos_ui.dart';
+
+class MacosTableHeader<T> extends StatelessWidget {
+  const MacosTableHeader({
```
Please add documentation here

GroovinChip commented on lib/src/layout/table/macos_table_header.dart:
```diff
@@ -0,0 +1,154 @@
+
+  static const double horizontalPadding = 10;
+
+  final List<ColumnDefinition<T>> columnDefinitions;
+  final Function(ColumnDefinition<T>)? columnHeaderClicked;
```
Please add documentation here

GroovinChip commented on lib/src/layout/table/macos_table_row.dart:
```diff
@@ -0,0 +1,143 @@
+import 'package:macos_ui/macos_ui.dart';
+import 'package:macos_ui/src/layout/table/macos_table_selection.dart';
+
+/// A widget to display a single row inside a [MacosTable].
```
```suggestion
/// {@template macosTableRow} 
/// A widget to display a single row inside a [MacosTable].
/// {@endtemplate}
```

GroovinChip commented on lib/src/layout/table/macos_table_row.dart:
```diff
@@ -0,0 +1,143 @@
+
+/// A widget to display a single row inside a [MacosTable].
+class MacosTableRow<T> extends StatelessWidget {
+  const MacosTableRow({
```
```suggestion
  /// {@macro macosTableRow}
  const MacosTableRow({
```

GroovinChip commented on lib/src/layout/table/macos_table_row.dart:
```diff
@@ -0,0 +1,143 @@
+    required this.data,
+  }) : hasEvenRowHighlight = index % 2 == 1;
+
+  /// A map from column index to its [TableColumnWidth].
+  /// The same information is also contained in the [columnDefinitions],
```
```suggestion
  /// A map from column index to its [TableColumnWidth].
  ///
  /// The same information is also contained in the [columnDefinitions],
```

GroovinChip commented on lib/src/layout/table/macos_table_row.dart:
```diff
@@ -0,0 +1,143 @@
+  /// The height of this row.
+  final double rowHeight;
+
+  /// All rows in a table are numbered, this is the index of this row.
+  /// Used to highlight every other row through [hasEvenRowHighlight].
```
```suggestion
  /// This is the index of this row in the table.
  ///
  /// Used to highlight every other row through [hasEvenRowHighlight].
```

GroovinChip commented on lib/src/layout/table/macos_table_selection.dart:
```diff
@@ -0,0 +1,31 @@
+import 'package:macos_ui/macos_ui.dart';
+import 'package:macos_ui/src/library.dart';
+
+/// The selection state of a [MacosTable].
```
```suggestion
/// {@template macosTableSelection}
/// The selection state of a [MacosTable].
/// {@endtemplate}
```

GroovinChip commented on lib/src/layout/table/macos_table_selection.dart:
```diff
@@ -0,0 +1,31 @@
+/// The selection state of a [MacosTable].
+@immutable
+class MacosTableSelection<T> {
+  // TODO: Implement selecting multiple rows.
```
```suggestion
  /// {@macro macosTableSelection}
  // TODO: Implement selecting multiple rows.
```

GroovinChip commented on lib/src/layout/table/macos_table_selection.dart:
```diff
@@ -0,0 +1,31 @@
+}
+
+/// Represents a change of a [MacosTable] selection.
+/// Includes the old and the new selection.
```
```suggestion
/// Represents a change of a [MacosTable] selection.
///
/// Includes the old and the new selection.
```

GroovinChip commented on lib/src/layout/table/macos_table_selection.dart:
```diff
@@ -0,0 +1,31 @@
+
+/// Represents a change of a [MacosTable] selection.
+/// Includes the old and the new selection.
+/// Used to figure out whether a row should rebuild after a selection change.
```
```suggestion
/// {@template macosTableSelectionChange}
/// Used to figure out whether a row should rebuild after a selection change.
/// {@endtemplate}
```

GroovinChip commented on lib/src/layout/table/macos_table_selection.dart:
```diff
@@ -0,0 +1,31 @@
+/// Used to figure out whether a row should rebuild after a selection change.
+@immutable
+class MacosTableSelectionChange {
+  const MacosTableSelectionChange(
```
```suggestion
  /// {@macro macosTableSelectionChange}
  const MacosTableSelectionChange(
```

GroovinChip commented on lib/src/layout/table/macos_table_view.dart:
```diff
@@ -0,0 +1,82 @@
+
+  @override
+  Widget build(BuildContext context) {
+    /// A map from column index to its [TableColumnWidth].
+    /// Every Table widget needs this, so it is created once and cached here.
```
```suggestion
    /// A map from column index to its [TableColumnWidth].
    ///
    /// Every Table widget needs this, so it is created once and cached here.
```

GroovinChip committed:
Update lib/src/layout/table/macos_table_columns.dart

GroovinChip committed:
Apply suggestions from code review

GroovinChip committed:
Apply suggestions from code review

GroovinChip reviewed (changes requested):
One more thing - can you please post a screenshot of the table in dark mode? I want to make sure that it looks good with dark theme.

Also can you please run `flutter format` on `macos_table_datasource.dart`

Yetispapa commented:
@pithuene looking forward for the merge! Can you maybe give me an example, how the ordering actually work?

It seems, when I change the datasource like so:

```

StreamBuilder(
          stream: widget.bloc.items,
          initialData: widget.bloc.items.value,
          builder: (BuildContext context, snapshot) {

....

    changeOrder: (order) {
      // this will update the StreamBuilder
      widget.bloc.changeOrder(order.direction.name);
      return true;
    },
```

The ordering gets messed up somehow. Can you help me here out. Thanks

pithuene commented:
I can take a look @Yetispapa but I would need more context, its hard to tell much just from the code you've shown.
You need to implement the ordering yourself (which you probably did in `bloc.changeOrder`, but I can't see).
Also I am not sure why you wrapped the table in a `StreamBuilder`, you should call `dataChanged` on the data source after the data has changed and let the table do the rebuilding internally.

Yetispapa commented:
@pithuene, thanks for the quick repsonse. Alright. Let me check it again.

GroovinChip commented:
@pithuene anything new to report on this?

pithuene commented:
@GroovinChip not yet. I've been too busy the last few weeks, might find the time next week though.

GroovinChip commented:
No problem my dude, take your time. Your life comes first :)

romanr commented:
PR gonna be 2 years in couple e months 🍾

pithuene commented:
Yeah, sorry about that, but I won't be able to finish it. If anyone wants to use my code and take it from here, feel free. But I'm closing the PR for now.
